src/kernel/scheduler.zig

```diff
@@ -0,0 +1,583 @@
+///
+/// Arguments:
+///     IN pid: u32 - The PID to free.
+///
+fn freePid(pid: u32) void {
+    all_pids.clearEntry(pid) catch panic(@errorReturnTrace(), "PID {} not allocated", .{pid});
+}
+
+///
+/// Round robin. This will pick the next task to be run from the queue. It will add the current
+/// task to the end of the queue and pop the next task from the front as set this as the current
+/// task. Interrupts are assumed disabled.
+///
+fn pickNextTask() void {
+    // Get the next task
+    if (tasks.pop()) |new_task_node| {
+        var next_task = new_task_node.data;
+        tasks.destroyNode(new_task_node, alloc);
+
+        // Now that we have the new task and deleted its node, there "will" be space for another
+        // so the allocation shouldn't fail as interrupts are disabled.
+        // So if there is a failure, then wow :P This will be the case for multiprocessor, but
+        // for a later time
+
+        // Put the current task back in the queue, at the back
+        var cur_elem = tasks.createNode(current_task, alloc) catch |e| {
+            panic(@errorReturnTrace(), "Error in pickNextTask, can't save previous task: {}\n", .{e});
+        };
+        tasks.prepend(cur_elem);
+
+        // Set the next task
+        current_task = next_task;
+    } else {
+        // Not more tasks, this isn't possible as the idle task is noreturn
+        panic(@errorReturnTrace(), "No more tasks?\n", .{});
+    }
+}
+
+///
+/// The idle task that just halts the CPU but the CPU can still handle interrupts.
+///
+fn idle() noreturn {
+    arch.spinWait();
+}
+
+///
+/// Create a new task and add it to the scheduling queue. No locking.
+///
+/// Arguments:
+///     IN comptime entry_point: var - The entry point into the task. This must be a function.
+///
+/// Error: std.mem.Allocator.Error
+///     OutOfMemory - If there isn't enough memory for the a task/stack. Any memory allocated will
+///                   be freed on return.
+///
+pub fn createNewTask(comptime entry_point: var) Allocator.Error!void {
+    // Ensure the entry point is a function
+    comptime assert(@typeInfo(@TypeOf(entry_point)) == .Fn);
+
+    var task = try Task.init(entry_point);
+    errdefer task.destroy();
+
+    var elem = try tasks.createNode(task, alloc);
+    tasks.prepend(elem);
+}
+
+///
+/// Schedule a task switch. This will first save the the current tasks stack pointer, then will
+/// pick the next task to be run and return the stack pointer for the next task to be loaded into
+/// ESP to load the next task stack to pop off its state. Interrupts are assumed disabled.
+///
+/// Argument:
+///     IN ctx: *arch.CpuState - Pointer to the exception context containing the contents
+///                              of the registers at the time of a exception.
+///
+/// Return: u32
+///     The new stack pointer to the next stack of the next task.
+///
+pub fn schedule(ctx: *arch.CpuState) u32 {
```

Comment:
In my mind "schedule" means put a task on the queue. Would you consider changing it to be something like "switchTask" with it taking the task to switch to, then anyone wanting to switch to the next would call "pickNextTask" then call this to perform the actual switch?

Reply:
Sure, so the RTC will call pickNextTask instead

Reply:
Yeah it would call pickNextTask to get the next task to run and then switchTask to do the actual switch.